### DISCUSSION TITLE
Broad types accepted by the starlite constructor.

### DISCUSSION THREAD

**Discussion by peterschutt** [litestar-org/litestar#1134] [Potential Issues]
As a very specific example, I have a question of the need for the application instance to receive anything other than a dict as the value for the `initial_state` parameter. By accepting `ImmutableState` and `tuple[str, Any]` it makes it a lot harder for plugins to use the app state if they need to.

```py
    def set_initial_state_in_hook(app_config: AppConfig) -> AppConfig:
        if isinstance(app_config.initial_state, ImmutableState):
            return
        if isinstance(app_config.initial_state, tuple):
            app_config.state = dict(app_config.state)
        app_config.initial_state["c"] = "D"
        app_config.initial_state["e"] = "f"
        return app_config
```

Handlers end up having to do lots of state checking before they can do anything.. and if `initial_state` is an instance of immutable state, then they can't really use it. Writing plugins where initial state may be immutable means that you cannot really use it downstream reliably (you can if the plugin is a personal thing, but as soon as you want to publish it as a library you can't).

We could have another app/config parameter `app_state_immutable` or `app_state_type: type[ImmutableState] = State` which could be used to indicate whether the application state that is constructed should be either `ImmutableState` or `State`, and I don't think it would be too unreasonable to ask users to pass dict(initial_state) to the constructor if they do have a sequence of k/v pairs.

Less specifically, we also accept a few parameters that can be either scalar, or a list. I get that this is something of a convenience, but we are only talking about asking users to provide `...=[handler]` instead of `...=handler`, which I'd be very surprised if we ever received complaints about. Again from a plugin perspective, before one can interact with any of the attributes that might be a list or a scalar, they need to use something like this (which is an actual helper I use in `starlite-saqlalchemy`):

```py
    @staticmethod
    def _ensure_list(item: list[T] | T) -> list[T]:
        if isinstance(item, list):
            return item
        return [] if item is None else [item]
```

So I'd like to gauge the community feelings about simplifying these params to the constructor for 2.0.

**Comment by provinzkraut** [comment]
I think it's a good idea to narrow the types down. It's also more consistent, since we do this in some parts already.

As for the state specifically, I think #1111 also plays into this. If you can use a custom state class, it would make a lot of sense that it can be used as an initial state object as well.

**Comment by Goldziher** [comment]
sure

**Comment by peterschutt** [comment]
OK I've done a pass over the app/router/controller/handler parameters in #1140 for 2.0.
